lib/syncthing/syncthing.go

```diff
@@ -77,6 +77,7 @@ type App struct {
 	stopOnce          sync.Once
 	mainServiceCancel context.CancelFunc
 	stopped           chan struct{}
+	Model             model.Model
```

Comment:
I am a bit late: I'd rather not expose the entirety of the model interface here. For example we definitely don't want anyone to use the suture service methods - there are dedicated methods to start/stop `App`. Lets make the model member internal and expose just the methods you need.

Reply:
Agreed, sorry.

Reply:
I agree that it is a good idea not to expose too much. We could design an interface/type that wraps Model and only exposes specific methods (the iOS app uses quite a lot, see [this search](https://github.com/search?q=repo%3Apixelspark%2Fsushitrain%20app.Model&type=code). 

We should however also consider who we are protecting from in the first place: the only users of this variable are accessing it either from Syncthing code itself (fair enough, protecting members makes for cleaner separation of concerns) or are third parties who import the packages (such as the iOS app) - these are ([as you rightfully warned me about already](https://forum.syncthing.net/t/idea-for-on-demand-syncing-on-ios-need-some-guidance-on-implementing/22408/2?u=pixelspark) :-)) basically doing something unsupported anyway.

Reply:
:+1: 

> Additionally we could even name the interface `Internal` and document that it's A) potentially dangerous (e.g. it's probably fine to use for whatever, but likely has performance implications and might also have unexpected effects) and B) API may change at any time without notice (or in short "here be dragons").

 B actually applies to the whole package I think (not sure what would be an appropriate place to put the 'here be dragons' notice for it). Labelling the extra special ones as 'Internal' is a good signal I think.

> Also just mentioning some "bigger" ideas, I am not asking for this to happen (now) - just in case that's something you'd fancy: One of the bigger users of model methods is your [server/file request method](https://github.com/pixelspark/sushitrain/blob/1b813760dc1b13d0e56b69e826ffbb2e1957f701/SushitrainCore/src/server.go#L122). That looks like a simplified method of what our puller does. So it feels like instead of using a bunch of model methods on your end (availability, requestglobal, ...), that could live in our code-base. Ideally re-using puller code, but that's somewhere between annoying and impractical, but otherwise just as a utility method in `lib/syncthing` to request (part of) a file. That way all those methods don't need to be exposed and you can still do the same thing in your server.

Yeah, agreed. It was easier for me to write it this way than to figure out how to integrate with the puller logic (some of which may also not be exposed yet?) and whether it could do throttling and range requests. But if this already exists in the codebase, I should re-use it.

Reply:
Yeah that's the thing: It's true that we consider everything internal and don't give any stability guarantees. It's just that I wont even think about changing an external method in lib/model, while I will in lib/syncthing - they have different purposes and "internalness" in my mind. Of course "in my mind" makes it not discoverable and non-canonical :) (1)

As for the concerns you mentioned around using puller code: Fully agreed, that's the kind of thing I meant by it not being practical. I think the more realistic option would be to create some kind of `RequestData` utility function in lib/model exposed through lib/syncthing which does almost the same ase your `/file` handler func, minus the http specific stuff (with or without throttling). Then you can use that function in your http server.

(1): Until last week that was mainly because I thought Martchus wrapper still did use it, but apparently he doesn't. Still the intent of the package is that still (external use), so I consider it as less internal/more stable than anything else. Especially model, that's probably the most "central chaotic bunch of code that glues everything together to make it work" part of syncthing :)

Reply:
@imsodin As for the wrapper, something like this? https://github.com/pixelspark/syncthing/commit/5dc100beaaced4108a2c4408b4cc22cecdfa60e0?diff=unified&w=0

(This has all the methods the iOS app is calling)

Reply:
> Something like this, but with the interface in lib/syncthing: The interface exists to define what to expose in syncthing.App, that's not a concern of lib/model.

Makes sense (only reason to put it in lib/model would be if the whole `Model` interface could be made private, but that seems complicated)

> Also this seems to have a lot of methods I didn't notice before, and e.g. `GlobalFileInfo` I can't find in your repo?

Hm, a call to `CurrentGlobalFile` has been in there all the time (I renamed it here to `GlobalFileInfo` to make it a bit more consistent). Did you click 'show X more matches' in the Git search?

<img width="200" alt="image" src="https://github.com/user-attachments/assets/51de99a9-ab05-4663-bdc1-a1cf6e74c77a">

Perhaps you could advise which methods (if any) you think are less ideal to expose. In some cases the scope of the exposed methods could be reduced. As an example, I am using `DBSnapshot`, but only to get some statistics, so I could also change the wrapper interface to be limited to just that and not expose DBSnapshot entirely.

Reply:
I just missed the rename... sorry.

There aren't any methods that stand out as problematic, I was just confused because apparently I missed a lot of model usages when I skimmed your code. As you mentioned it: Your usage of DBSnapshot looks like folder summaries, you might just be able to use `model.FolderSummaryService` that's already exported? Otherwise I think it's fine like this - when you PR I'll take a pass at naming/params, maybe something comes up, but otherwise this seems like a good start.

Reply:
@imsodin let's continue here: https://github.com/syncthing/syncthing/pull/9627